File: src/service/appointment.service.ts
import { prisma } from "../utils/prisma";
import { CreateSlotInput, BookAppointmentInput, UpdateAppointmentStatusInput } from "../validators/appointment.validator";
import { BadRequestError, NotFoundError, UnauthorizedError } from "../utils/errors";
import { AppointmentStatus, RelationshipStatus } from "@prisma/client";

export class AppointmentService {
  // ==========================================
  // TEACHER METHODS
  // ==========================================

  static async createSlot(teacherId: string, data: CreateSlotInput) {
    const dateObj = new Date(data.date);
    if (isNaN(dateObj.getTime())) {
      throw new BadRequestError("Invalid date.");
    }

    return prisma.appointmentSlot.create({
      data: {
        teacherId,
        date: dateObj,
        startTime: data.startTime,
        endTime: data.endTime,
      },
    });
  }

  static async closeSlot(slotId: string, teacherId: string) {
    const slot = await prisma.appointmentSlot.findUnique({ where: { id: slotId } });
    if (!slot) throw new NotFoundError("Appointment slot not found.");
    if (slot.teacherId !== teacherId) throw new UnauthorizedError("You do not own this slot.");

    return prisma.appointmentSlot.update({
      where: { id: slotId },
      data: { isClosed: true },
    });
  }

  static async getTeacherAppointments(teacherId: string) {
    return prisma.appointment.findMany({
      where: {
        slot: {
          teacherId,
        },
      },
      include: {
        slot: true,
        parent: { select: { id: true, firstName: true, lastName: true, email: true, phoneNumber: true } },
        student: { select: { id: true, firstName: true, lastName: true } },
      },
      orderBy: {
        slot: { date: "asc" },
      },
    });
  }

  static async updateAppointmentStatus(appointmentId: string, teacherId: string, status: AppointmentStatus) {
    const appointment = await prisma.appointment.findUnique({
      where: { id: appointmentId },
      include: { slot: true },
    });

    if (!appointment) throw new NotFoundError("Appointment not found.");
    if (appointment.slot.teacherId !== teacherId) throw new UnauthorizedError("You do not own this appointment.");

    return prisma.appointment.update({
      where: { id: appointmentId },
      data: { status },
    });
  }

  // ==========================================
  // PARENT METHODS
  // ==========================================

  static async getAvailableSlots(teacherId: string) {
    // A slot is available if it is not closed and has less than 2 active appointments
    const slots = await prisma.appointmentSlot.findMany({
      where: {
        teacherId,
        isClosed: false,
        date: { gte: new Date() }, // Only future or today slots
      },
      include: {
        appointments: {
          where: {
            status: { in: [AppointmentStatus.PENDING, AppointmentStatus.ACCEPTED] },
          },
        },
      },
      orderBy: [
        { date: "asc" },
        { startTime: "asc" }
      ]
    });

    // Filter slots by capacity (max 2)
    return slots.filter((slot) => slot.appointments.length < 2).map(slot => {
      const { appointments, ...slotData } = slot;
      return {
        ...slotData,
        availableCapacity: 2 - appointments.length
      };
    });
  }

  static async bookAppointment(parentId: string, data: BookAppointmentInput) {
    const slot = await prisma.appointmentSlot.findUnique({
      where: { id: data.slotId },
      include: {
        appointments: {
          where: {
            status: { in: [AppointmentStatus.PENDING, AppointmentStatus.ACCEPTED] },
          },
        },
      },
    });

    if (!slot) throw new NotFoundError("Appointment slot not found.");
    if (slot.isClosed) throw new BadRequestError("This slot is closed.");
    if (slot.appointments.length >= 2) throw new BadRequestError("This slot is fully booked.");

    // Verify parent-student relationship
    const relationship = await prisma.parentStudentRelationship.findFirst({
      where: {
        parentId,
        studentId: data.studentId,
        status: RelationshipStatus.APPROVED,
      },
    });

    if (!relationship) {
      throw new UnauthorizedError("You do not have an approved relationship with this student.");
    }

    // Verify teacher-student relationship
    const assignment = await prisma.teacherClassSubject.findFirst({
      where: {
        teacherId: slot.teacherId,
        class: {
          students: {
            some: { studentId: data.studentId }
          }
        }
      },
    });

    if (!assignment) {
      throw new UnauthorizedError("This teacher does not teach this student.");
    }

    // Check if this specific parent already booked this slot for this student
    const existingBooking = slot.appointments.find(
      (appt) => appt.parentId === parentId && appt.studentId === data.studentId
    );
    if (existingBooking) {
      throw new BadRequestError("You have already booked this slot for this student.");
    }

    return prisma.appointment.create({
      data: {
        slotId: data.slotId,
        parentId,
        studentId: data.studentId,
        reason: data.reason,
      },
      include: {
        slot: true,
      }
    });
  }

  static async cancelAppointment(appointmentId: string, parentId: string) {
    const appointment = await prisma.appointment.findUnique({
      where: { id: appointmentId },
    });

    if (!appointment) throw new NotFoundError("Appointment not found.");
    if (appointment.parentId !== parentId) throw new UnauthorizedError("You do not own this appointment.");
    if (appointment.status === AppointmentStatus.CANCELLED_BY_PARENT) {
       throw new BadRequestError("Appointment is already cancelled.");
    }

    return prisma.appointment.update({
      where: { id: appointmentId },
      data: { status: AppointmentStatus.CANCELLED_BY_PARENT },
    });
  }

  static async getParentAppointments(parentId: string) {
    return prisma.appointment.findMany({
      where: { parentId },
      include: {
        slot: {
          include: {
            teacher: { select: { id: true, firstName: true, lastName: true } }
          }
        },
        student: { select: { id: true, firstName: true, lastName: true } },
      },
      orderBy: {
        createdAt: "desc",
      },
    });
  }
}

File: lib/services/appointment_service.dart
import '../core/api_client.dart';

class AppointmentSlotItem {
  final String id;
  final String date;
  final String startTime;
  final String endTime;
  final bool isClosed;
  final int availableCapacity;

  AppointmentSlotItem({
    required this.id,
    required this.date,
    required this.startTime,
    required this.endTime,
    this.isClosed = false,
    this.availableCapacity = 2,
  });

  factory AppointmentSlotItem.fromJson(Map<String, dynamic> json) {
    return AppointmentSlotItem(
      id: json['id'] as String? ?? '',
      date: json['date'] as String? ?? '',
      startTime: json['startTime'] as String? ?? '',
      endTime: json['endTime'] as String? ?? '',
      isClosed: json['isClosed'] as bool? ?? false,
      availableCapacity: (json['availableCapacity'] as num?)?.toInt() ?? 2,
    );
  }
}

class AppointmentItem {
  final String id;
  final String parentName;
  final String parentEmail;
  final String studentName;
  final String teacherName;
  final String date;
  final String startTime;
  final String endTime;
  final String reason;
  final String status;

  AppointmentItem({
    required this.id,
    required this.parentName,
    required this.parentEmail,
    required this.studentName,
    required this.teacherName,
    required this.date,
    required this.startTime,
    required this.endTime,
    required this.reason,
    required this.status,
  });

  factory AppointmentItem.fromJson(Map<String, dynamic> json) {
    String pName = 'Parent';
    String pEmail = '';
    if (json['parent'] != null && json['parent'] is Map<String, dynamic>) {
      final p = json['parent'] as Map<String, dynamic>;
      pName = '${p['firstName'] ?? ''} ${p['lastName'] ?? ''}'.trim();
      pEmail = p['email'] as String? ?? '';
    }

    String sName = 'Student';
    if (json['student'] != null && json['student'] is Map<String, dynamic>) {
      final s = json['student'] as Map<String, dynamic>;
      sName = '${s['firstName'] ?? ''} ${s['lastName'] ?? ''}'.trim();
    }

    String tName = 'Teacher';
    String dt = '';
    String st = '';
    String et = '';
    if (json['slot'] != null && json['slot'] is Map<String, dynamic>) {
      final sl = json['slot'] as Map<String, dynamic>;
      dt = sl['date'] as String? ?? '';
      st = sl['startTime'] as String? ?? '';
      et = sl['endTime'] as String? ?? '';
      if (sl['teacher'] != null && sl['teacher'] is Map<String, dynamic>) {
        final t = sl['teacher'] as Map<String, dynamic>;
        tName = '${t['firstName'] ?? ''} ${t['lastName'] ?? ''}'.trim();
      }
    }

    return AppointmentItem(
      id: json['id'] as String? ?? 'apt-${DateTime.now().millisecondsSinceEpoch}',
      parentName: pName.isNotEmpty ? pName : 'Parent',
      parentEmail: pEmail,
      studentName: sName.isNotEmpty ? sName : 'Student',
      teacherName: tName.isNotEmpty ? tName : 'Teacher',
      date: dt.isNotEmpty ? dt : (json['date'] as String? ?? ''),
      startTime: st.isNotEmpty ? st : (json['startTime'] as String? ?? ''),
      endTime: et.isNotEmpty ? et : (json['endTime'] as String? ?? ''),
      reason: json['reason'] as String? ?? 'General Conference Request',
      status: json['status'] as String? ?? 'PENDING',
    );
  }
}

class AppointmentService {
  AppointmentService({ApiClient? apiClient}) : _apiClient = apiClient ?? ApiClient();

  final ApiClient _apiClient;

  Future<List<AppointmentItem>> getTeacherAppointments() async {
    try {
      final res = await _apiClient.get('/appointments/teacher', requireAuth: true);
      final list = (res['data'] as List<dynamic>?) ?? [];
      return list.map((e) => AppointmentItem.fromJson(e as Map<String, dynamic>)).toList();
    } catch (_) {
      return [];
    }
  }

  Future<List<AppointmentItem>> getParentAppointments() async {
    try {
      final res = await _apiClient.get('/appointments/parent', requireAuth: true);
      final list = (res['data'] as List<dynamic>?) ?? [];
      return list.map((e) => AppointmentItem.fromJson(e as Map<String, dynamic>)).toList();
    } catch (_) {
      return [];
    }
  }

  Future<List<AppointmentSlotItem>> getAvailableSlots(String teacherId) async {
    try {
      final res = await _apiClient.get('/appointments/teachers/$teacherId/slots', requireAuth: true);
      final list = (res['data'] as List<dynamic>?) ?? [];
      return list.map((e) => AppointmentSlotItem.fromJson(e as Map<String, dynamic>)).toList();
    } catch (_) {
      return [];
    }
  }

  Future<bool> createSlot({
    required String date,
    required String startTime,
    required String endTime,
  }) async {
    try {
      await _apiClient.post(
        '/appointments/slots',
        body: {'date': date, 'startTime': startTime, 'endTime': endTime},
        requireAuth: true,
      );
      return true;
    } catch (_) {
      return false;
    }
  }

  Future<bool> bookAppointment({
    required String slotId,
    required String studentId,
    String? reason,
  }) async {
    try {
      await _apiClient.post(
        '/appointments/book',
        body: {
          'slotId': slotId,
          'studentId': studentId,
          if (reason != null && reason.isNotEmpty) 'reason': reason,
        },
        requireAuth: true,
      );
      return true;
    } catch (_) {
      return false;
    }
  }

  Future<bool> cancelAppointment(String appointmentId) async {
    try {
      await _apiClient.patch(
        '/appointments/$appointmentId/cancel',
        body: {},
        requireAuth: true,
      );
      return true;
    } catch (_) {
      return false;
    }
  }

  Future<bool> updateAppointmentStatus(String appointmentId, String status) async {
    try {
      await _apiClient.patch(
        '/appointments/$appointmentId/status',
        body: {'status': status},
        requireAuth: true,
      );
      return true;
    } catch (_) {
      return false;
    }
  }
}
